app/Http/Controllers/ContestEntriesController.php

```diff
@@ -5,20 +5,142 @@
 
 namespace App\Http\Controllers;
 
+use App\Exceptions\InvariantException;
 use App\Models\Contest;
 use App\Models\ContestEntry;
+use App\Models\ContestJudgeScore;
+use App\Models\ContestJudgeVote;
 use App\Models\UserContestEntry;
 use Auth;
+use Illuminate\Support\Facades\DB;
 use Request;
 
 class ContestEntriesController extends Controller
 {
+    public function judgeResults($id)
+    {
+        $entry = ContestEntry::with('contest')
+            ->with('contest.entries')
+            ->with('contest.scoringCategories')
+            ->with('judgeVotes')
+            ->with('judgeVotes.scores')
+            ->with('judgeVotes.user')
+            ->with('judgeVotes.scores.category')
+            ->with('user')
+            ->withSum('scores', 'value')
+            ->findOrFail($id);
+
+        abort_if(!$entry->contest->isJudged() || !$entry->contest->show_votes, 404);
+
+        $contestJson = json_item(
+            $entry->contest->loadSum('scoringCategories', 'max_value'),
+            'Contest',
+            ['max_judging_score']
+        );
+
+        $entryJson = json_item($entry, 'ContestEntry', [
+            'judge_votes.user',
+            'judge_votes.total_score',
+            'judge_votes.scores.category',
+            'results',
+            'user',
+        ]);
+
+        $entriesJson = json_collection($entry->contest->entries, 'ContestEntry');
+
+        return ext_view('contest_entries.judge-results', [
+            'contestJson' => $contestJson,
+            'entryJson' => $entryJson,
+            'entriesJson' => $entriesJson,
+        ]);
+    }
+
+    public function judgeVote($id)
+    {
+        $entry = ContestEntry::with('contest')
+            ->with('contest.judges')
+            ->with('contest.scoringCategories')
+            ->with('judgeVotes')
+            ->findOrFail($id);
+
+        priv_check('ContestJudge', $entry->contest)->ensureCan();
+
+        // so that admin can't submit vote if not judge
+        abort_if(!$entry->contest->isJudge(auth()->user()), 403);
+
+        $params = get_params(request()->all(), null, [
+            'scores:array',
+            'comment',
+        ]);
+
+        $scores = collect($params['scores']);
+        $comment = $params['comment'];
+
+        DB::transaction(function () use ($scores, $comment, $entry) {
+            $vote = $entry->judgeVotes->where('user_id', auth()->user()->getKey())->first();
+
+            if ($vote !== null) {
+                if ($comment !== $vote->comment) {
+                    $vote->update(['comment' => $comment]);
+                }
+            } else {
+                $vote = ContestJudgeVote::create([
+                    'comment' => $comment,
+                    'contest_entry_id' => $entry->getKey(),
+                    'user_id' => auth()->user()->getKey(),
+                ]);
+            }
+
+            foreach ($entry->contest->scoringCategories as $category) {
+                $score = $scores
```

Comment:
Can use `Ds\Set` and get the intersection/diff for missing ids.
Or, just default missing values to 0.